mne/io/eeglab/eeglab.py

```diff
@@ -0,0 +1,140 @@
+from ..base import _BaseRaw
+from ..constants import FIFF
+
+
+def _topo_to_sph(theta, radius):
+    """Convert 2D topo coordinates to spherical.
+    """
+    sph_phi = (0.5 - radius) * 180
+    sph_theta = -theta
+    return sph_phi, sph_theta
+
+
+def _sph_to_cart(azimuth, elevation, r):
+    """Convert spherical to cartesian coordinates.
+    """
+    x = r * np.cos(elevation) * np.cos(azimuth)
+    y = r * np.cos(elevation) * np.sin(azimuth)
+    z = r * np.sin(elevation)
+    return x, y, z
+
+
+def _get_info(eeg, eog, ch_fname):
+    """Get measurement info.
+    """
+    if not ch_fname.endswith('.locs'):
+        msg = """Currently, only the .locs file format is supported.
+              Please contact mne-python developers for more information"""
+        raise NotImplementedError(msg)
+
+    ch_names = np.loadtxt(ch_fname, dtype='S4', usecols=[3]).tolist()
+    info = create_info(sfreq=eeg.srate, ch_names=ch_names)
+
+    # load channel locations
+    dtype = {'names': ('angle', 'radius'), 'formats': ('f4', 'f4')}
+    angle, radius = np.loadtxt(ch_fname, dtype=dtype, usecols=[1, 2],
+                               unpack=True)
+    sph_phi, sph_theta = _topo_to_sph(angle, radius)
+    sph_radius = np.ones((eeg.nbchan, ))
+    x, y, z = _sph_to_cart(sph_theta / 180 * np.pi,
+                           sph_phi / 180 * np.pi, sph_radius)
+    if eog is None:
+        eog = [idx for idx, ch in enumerate(ch_names) if ch.startswith('EOG')]
+
+    for idx, ch in enumerate(info['chs']):
+        ch['loc'][:3] = np.array([x[idx], y[idx], z[idx]])
+        ch['unit'] = FIFF.FIFF_UNIT_V
+        ch['coord_frame'] = FIFF.FIFFV_COORD_HEAD,
+        ch['coil_type'] = FIFF.FIFFV_COIL_EEG,
+        ch['kind'] = FIFF.FIFFV_EEG_CH
+
+        if ch['ch_name'] in eog:
+            ch['coil_type'] = FIFF.FIFFV_COIL_NONE
+            ch['kind'] = FIFF.FIFFV_EOG_CH
+
+    return info
+
+
+def read_raw_set(fname, ch_fname, eog=None, verbose=None):
+    """Read an EEGLAB .set file
+
+    Parameters
+    ----------
+    fname : str
+        Path to the .set file.
+    ch_fname : str
+        Path to the file containing channel locations. Currently,
+        only .loc extension is supported.
+    eog : list or tuple
+        Names of channels or list of indices that should be designated
+        EOG channels. If None (default), the channel names beginning with
+        ``EOG`` are used.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    raw : Instance of RawSet
+        A Raw object containing EEGLAB .set data.
+    """
+    return RawSet(fname, ch_fname, eog, verbose)
+
+
+class RawSet(_BaseRaw):
+    """Raw object from EEGLAB .set file.
+
+    Parameters
+    ----------
+    fname : str
+        Path to the .set file.
+    ch_fname : str
+        Path to the file containing channel locations. Currently,
+        only .loc extension is supported.
```

Comment:
On Mon, Dec 7, 2015 at 1:46 PM, jona-sassenhagen <notifications@github.com>
wrote:

> In mne/io/eeglab/eeglab.py
> <https://github.com/mne-tools/mne-python/pull/2676#discussion_r46816660>:
>
> > +    raw : Instance of RawSet
> > +        A Raw object containing EEGLAB .set data.
> > +    """
> > +    return RawSet(fname, ch_fname, eog, verbose)
> > +
> > +
> > +class RawSet(_BaseRaw):
> > +    """Raw object from EEGLAB .set file.
> > +
> > +    Parameters
> > +    ----------
> > +    fname : str
> > +        Path to the .set file.
> > +    ch_fname : str
> > +        Path to the file containing channel locations. Currently,
> > +        only .loc extension is supported.
>
> Why not use the general montage method that's used in the other import
> methods, and allows reading a bunch of other file formats?
>
good point. Done :) Now, the general montage method can also read .loc
files.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2676/files#r46816660>.
>